### PR TITLE
Fix leak and corruption bugs in CompositeByteBuf

### PR DESCRIPTION
**Pull request by njhill**
Motivation:

I came across two bugs while refactoring:
- Components removed due to capacity reduction aren't released
- Offsets aren't set correctly on empty components that are added between existing components, resulting in inconsistent/corrupt state of the buffer

Modifications:

Add unit tests which expose these bugs, fix them.

Result:

Bugs are fixed

### CODE REVIEW

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java [hunk 1]
```diff
@@ -1117,6 +1144,28 @@ public void testReleasesItsComponents() {
         assertEquals(0, buffer.refCnt());
     }
 
+    @Test
+    public void testReleasesOnShrink() {
+
+        ByteBuf b1 = PooledByteBufAllocator.DEFAULT.buffer(2).writeShort(1);
+        ByteBuf b2 = PooledByteBufAllocator.DEFAULT.buffer(2).writeShort(2);
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java]: Use Unpooled

File: buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java [hunk 2]
```diff
@@ -1117,6 +1144,28 @@ public void testReleasesItsComponents() {
...
+
+        // composite takes ownership of s1 and s2
+        ByteBuf composite = PooledByteBufAllocator.DEFAULT.compositeBuffer()
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java]: Use Unpooled

File: buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java [hunk 3]
```diff
@@ -1117,6 +1144,28 @@ public void testReleasesItsComponents() {
...
+            .addComponents(b1, b2);
+
+        assertEquals(4, composite.capacity());
+
+        composite.capacity(2);
+        assertEquals(2, composite.capacity());
+
+        // releasing composite should release the components
+        ReferenceCountUtil.release(composite);
```
Comment (normanmaurer) [inline buffer/src/test/java/io/netty/buffer/AbstractCompositeByteBufTest.java]: Just use composite.release() ?

**Comment by njhill** [comment]
@normanmaurer I've made the requested changes to the tests; I had just copied/modified existing tests, and also had an incorrect assumption in my head that unpooled buffers didn't do ref-counting!

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@njhill thanks a lot!

**Pull request merged**